### PR TITLE
gives host extra control over when the cycle should restart

### PR DESCRIPTION
**Pull request by 6fears7**
https://github.com/user-attachments/assets/4aca7bfc-1897-4cb0-9948-0bf3a411e608

### CODE REVIEW

File: Story/StoryHooks.cs [hunk 1]
```diff
@@ -89,8 +89,32 @@ private void StoryHooks()
...
+                }
+            }
+        }
+
+        private void PauseMenu_SpawnExitContinueButtons(On.Menu.PauseMenu.orig_SpawnExitContinueButtons orig, Menu.PauseMenu self)
+        {
+            orig(self);
+            if (isStoryMode(out var story) && story.wantsToRestartCycle)
+            {
+                var restartCycleButton = new Menu.SimpleButton(self, self.pages[0], self.Translate("Restart Cycle"), "ONLINERESTARTCYCLE", new Vector2(self.exitButton.pos.x - (self.continueButton.pos.x - self.exitButton.pos.x) - self.moveLeft - self.manager.rainWorld.options.SafeScreenOffset.x, Mathf.Max(self.manager.rainWorld.options.SafeScreenOffset.y, 15f)), new Vector2(110f, 30f));
```
Comment (youbitchoc) [inline Story/StoryHooks.cs:114]: we can use a SimplerButton callback instead and avoid hooking PauseMenu_Singal

File: Story/StoryHooks.cs [hunk 2]
```diff
@@ -249,6 +272,10 @@ private void TextPrompt_Update(On.HUD.TextPrompt.orig_Update orig, TextPrompt se
                     if (touchedInput)
                     {
                         self.gameOverMode = false;
+                        if (OnlineManager.lobby.isOwner)
+                        {
+                            story.wantsToRestartCycle = true;
```
Comment (youbitchoc) [inline Story/StoryHooks.cs:277]: also this variable name is misleading, imo `showRestartButton` would be a better name

File: Story/StoryHooks.cs (hunk 1 above)
Comment (youbitchoc) [reply Story/StoryHooks.cs:114]: also name the button something shorter like "Restart"

File: GameModes/StoryGameMode.cs [hunk 3]
```diff
@@ -26,6 +26,7 @@ public class StoryGameMode : OnlineGameMode
 
         public string? myLastDenPos = null;
         public bool hasSheltered = false;
+        public bool wantsToRestartCycle = false;
```
Comment (youbitchoc) [inline GameModes/StoryGameMode.cs:29]: do we _need_ to put this here? why not a static bool in StoryHooks?

**Review by youbitchoc — CHANGES_REQUESTED** [review]

File: Story/StoryHooks.cs [hunk 4]
```diff
@@ -89,9 +91,26 @@ private void StoryHooks()
...
+
+        private void PauseMenu_SpawnExitContinueButtons(On.Menu.PauseMenu.orig_SpawnExitContinueButtons orig, Menu.PauseMenu self)
+        {
+            orig(self);
+            if (isStoryMode(out var story) && showRestartButton)
+            {
+                var restartCycle = new SimplerButton(self, self.pages[0], self.Translate("Restart"), new Vector2(self.exitButton.pos.x - (self.continueButton.pos.x - self.exitButton.pos.x) - self.moveLeft - self.manager.rainWorld.options.SafeScreenOffset.x, Mathf.Max(self.manager.rainWorld.options.SafeScreenOffset.y, 15f)), new Vector2(110f, 30f));
+                restartCycle.OnClick += (_) =>
+                {
+                    showRestartButton = false;
```
Comment (youbitchoc) [inline Story/StoryHooks.cs:107]: can't rely on it getting unset from clicking, we'll have to unset it at game start or something like that

Comment (youbitchoc) [reply Story/StoryHooks.cs:107]: ... that won't get unset if you're revived though

imo fixing that is way too much hassle, thoughts on always adding the restart button for host? then we can get rid of `showRestartButton`

File: Story/StoryHooks.cs [hunk 5]
```diff
@@ -89,9 +91,26 @@ private void StoryHooks()
...
         }
 
 
+
+        private void PauseMenu_SpawnExitContinueButtons(On.Menu.PauseMenu.orig_SpawnExitContinueButtons orig, Menu.PauseMenu self)
+        {
+            orig(self);
+            if (isStoryMode(out var story) && showRestartButton)
+            {
+                var restartCycle = new SimplerButton(self, self.pages[0], self.Translate("Restart"), new Vector2(self.exitButton.pos.x - (self.continueButton.pos.x - self.exitButton.pos.x) - self.moveLeft - self.manager.rainWorld.options.SafeScreenOffset.x, Mathf.Max(self.manager.rainWorld.options.SafeScreenOffset.y, 15f)), new Vector2(110f, 30f));
```
Comment (youbitchoc) [inline Story/StoryHooks.cs:104]: also this variable should be named restartButton for consistency

**Review by youbitchoc — CHANGES_REQUESTED** [review]

File: Story/StoryHooks.cs (hunk 4 above)
Comment (6fears7) [reply Story/StoryHooks.cs:107]: You don't have to twist my arm.

File: GameModes/StoryGameMode.cs [hunk 6]
```diff
@@ -39,7 +39,6 @@ public void Sanitize()
             region = null;
             saveStateString = null;
             storyClientData?.Sanitize();
-
```
Comment (youbitchoc) [inline GameModes/StoryGameMode.cs]: 🙏

**Review by youbitchoc — DISMISSED** [review]
lgtm

**Comment by youbitchoc** [comment]
feel free to merge when tested

**Comment by 6fears7** [comment]
Tested, updated string to capitalization for consistency in button texts. Merging.

**Pull request merged**